[Haiku] Activate native backend

Sylvain78 opened the pull request:
- Activate native backend
- Look for getentropy in libbsd
- Prevents to use issetugid() as TODO is unresolved in Haiku source code.

When running the testsuite, those three tests failed :

- tests/lib-runtime-events/test_create_cursor_failures
- tests/lib-sys/signal
- tests/lib-unix/unix-socket/recvfrom_unix

dra27 commented:
It looks like you've generated `configure` by manually running `autoconf` - could you rebase the commits but use `tools/autogen` instead, please? (I'm happy to do it, if you prefer). This should fix both the Hygiene check and, more importantly, the Windows build.

Sylvain78 commented:
> It looks like you've generated `configure` by manually running `autoconf` - could you rebase the commits but use `tools/autogen` instead, please? (I'm happy to do it, if you prefer). This should fix both the Hygiene check and, more importantly, the Windows build.

Yes, please could you do the rebase and regenerate configure for me.

kit-ty-kate commented:
I've tested this change on a local Haiku install and encountered no problem so far (beyond the haiku specific things that required to set `./configure --prefix=/boot/home/config/non-packaged`)

Sylvain78 commented:
> I've tested this change on a local Haiku install and encountered no problem so far (beyond the haiku specific things that required to set `./configure --prefix=/boot/home/config/non-packaged`)

Did you launch the testsuite ? There must be 3/4 tests which don't pass.

kit-ty-kate commented:
> Did you launch the testsuite ? There must be 3/4 tests which don't pass.

I haven't yet, but it's running now, i'll update when it's done. I was able to compile and use opam though, so that amount of failures seems surprising to me.

kit-ty-kate commented:
ok the testsuite does partially fail but on things that one would expect to fail on such a platform (threads, float format, signals, …).

I've rebased your branch on trunk and regenerated the configure script in https://github.com/ocaml/ocaml/compare/trunk...kit-ty-kate:ocaml:haiku. Feel free to use the content of this branch

kit-ty-kate reviewed (approved):
as far as my limited testing on Haiku goes, this looks fine to me once the branch is rebased and the changelog added

Sylvain78 commented:
> as far as my limited testing on Haiku goes, this looks fine to me once the branch is rebased and the changelog added

done. Rebased and added a Changes entry.

kit-ty-kate commented:
> Rebased and added a Changes entry.

I think your `git push --force-with-lease` may have failed

dra27 commented on runtime/unix.c:
```diff
@@ -409,11 +409,12 @@ char *caml_secure_getenv (char const *var)
   return __secure_getenv (var);
-#elif defined(HAS_ISSETUGID)
+// TO RESTORE when TODO in haiku/src/libs/bsd/issetugid.c will be solved
+#elif defined(HAS_ISSETUGID) && !defined(__HAIKU__)
```
Surely this logic should be in `configure.ac` - i.e. we don't define `HAS_ISSETUGID` on Haiku, with the RESTORE note there?

dra27 reviewed (commented):
Sorry, I've been meaning to find some time to look at this, but not managing to, so just surfacing my opinion on merging: the testsuite isn't optional, even for a work-in-progress or partially-supported platform. The tests which don't pass ought to be marked as not-passing (i.e. adding a `not-haiku` or some such action to ocamltest, preferably with a note as to why in each test using it). My reasoning is that if someone else picks up the port and tries to use it, a _failing test_ looks like a bug (which might get then reported) vs a _skipped test_ or _missing feature_ which implies known work that needs to be done.

I'm very happy to help out with that at some point, but I'm afraid it comes under "hobby work" for me! 🙂

Sylvain78 commented:
All your remarks have been implemented. I add to replace addrlen by sizeof(struct sockaddr_un) in caml_unix_sendto(), it seems it is the right way to do , according to [https://pubs.opengroup.org/onlinepubs/9799919799/](https://pubs.opengroup.org/onlinepubs/9799919799/)

kit-ty-kate commented on testsuite/tests/lib-runtime-events/test_create_cursor_failures.ml:
```diff
@@ -2,6 +2,7 @@
  include unix;
  include runtime_events;
  hasunix;
+ not-beos_haiku;
```
I've also hit a failure here when testing https://github.com/ocaml/ocaml/pull/14431, however i think it would be more elegantly fixed by adding a way to detect if the current user is a super-user.

The issue is that the test does `chmod 000 <some file>` expecting the file to be unreadable but `root` on any Unix compatible system will simply read it regardless.

kit-ty-kate replied:
I've done that in https://github.com/ocaml/ocaml/pull/14435

kit-ty-kate replied:
https://github.com/ocaml/ocaml/pull/14435 has now been merged so this and its definition(s) can now be removed
```suggestion
```

Begasus commented:
I've done some local package on latest release with only enabling native backend in the configure script, so without this PR (will do that shortly and work together with @Sylvain78 to sort this out).
On running the tests here I got the following results
```
#	List of failed tests:
#    tests/lib-format/mc_pr586_par.ml
#    tests/lib-format/mc_pr586_par2.ml
#    tests/lib-runtime-events/test_create_cursor_failures.ml
#    tests/lib-sys/signal.ml
#    tests/lib-unix/unix-socket/recvfrom_unix.ml

#Summary:
#  1449 tests passed
#    57 tests skipped
#     5 tests failed
#     0 tests not started (parent test skipped or failed)
#     0 unexpected errors
#  1511 tests considered
```

kit-ty-kate commented on ocamltest/builtin_actions.ml:
```diff
@@ -176,6 +176,22 @@ let macos = make
+  (Actions_helpers.pass_or_skip (Ocamltest_config.system <> beos_system)
+    "not running on BeOS/Haiku"
+    "running on BeOS/Haiku")
+
```
this isn't used anymore
